src/hosting/runTags.ts

```diff
@@ -0,0 +1,183 @@
+import * as run from "../gcp/run";
+import * as api from "./api";
+import { FirebaseError } from "../error";
+import { flattenArray } from "../functional";
+
+/**
+ * Sentinel to be used when creating an api.Rewrite with the tag option but
+ * you don't yet know the tag. Resolve this tag by passing the rewrite into
+ * setRewriteTags
+ */
+export const TODO_TAG_NAME = "this is an invalid tag name so it cannot be real";
+
+/**
+ * Looks up all valid Hosting tags in this project and removes traffic targets
+ * from passed in services that don't match a valid tag.
+ * This makes no actual server-side changes to these services; you must then
+ * call run.updateService to save these changes. We divide this responsiblity
+ * because we want to possibly insert a new tagged target before saving.
+ */
+export async function gcTagsForServices(project: string, services: run.Service[]): Promise<void> {
+  // region -> service -> tags
+  // We cannot simplify this into a single map because we might be mixing project
+  // id and number.
+  const validTags: Record<string, Record<string, Set<string>>> = {};
```

Comment:
Fun pattern I use to carry your comment of "region -> service -> tags" through to code for readability is to name this `validTagsByServiceByRegion`. Then I know `validTagsByServiceByRegion[region][service] = tag` without any questions or referring to other comments. (Long because it's nested, but I think it _is_ nice!)